src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -169,20 +173,57 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
 
-                        _requestCompletionState = StreamCompletionState.Failed;
+                        if (_requestBodyAbandoned)
+                        {
+                            // See comments on _requestBodyAbandoned.
+                            // In this case, the request is still considered successful and we do not want to send a RST_STREAM, 
+                            // and we also don't want to propagate any error to the caller, in particular for non-duplex scenarios.
+                            Debug.Assert(_responseCompletionState == StreamCompletionState.Completed);
+                            _requestCompletionState = StreamCompletionState.Completed;
+                            Complete();
+                            return;
+                        }
+                        else
+                        {
+                            _requestCompletionState = StreamCompletionState.Failed;
 
-                        // Cancel above should ensure that the response is either Completed or Failed now.
-                        Debug.Assert(_responseCompletionState != StreamCompletionState.InProgress);
+                            // This duplicates some of the existing logic in Cancel.
+                            // Cancel takes the lock, so we can't call it here while holding the lock.
+                            // We can't call Cancel before taking the lock, because we don't want to cancel in the _requestBodyAbandoned case.
+                            // But we also don't want to call Cancel after releasing the lock, because that can cause a race
+                            // where the client sees the request body has failed but the response body hasn't.
+                            // We should consider how to refactor this logic to avoid code duplication.
 
-                        Reset();
+                            // Cancel the response body, if not complete.
+                            if (_responseCompletionState == StreamCompletionState.InProgress)
+                            {
+                                _responseCompletionState = StreamCompletionState.Failed;
+                            }
+
+                            // Discard any remaining buffered response data
+                            if (_responseBuffer.ActiveLength != 0)
+                            {
+                                _responseBuffer.Discard(_responseBuffer.ActiveLength);
+                            }
+
+                            _responseProtocolState = ResponseProtocolState.Aborted;
+
+                            signalWaiter = _hasWaiter;
+                            _hasWaiter = false;
+
+                            Reset();
+                        }
+                    }
+
+                    if (signalWaiter)
+                    {
+                        _waitSource.SetResult(true);
```

Comment:
Why don't we want to do this in the _requestBodyAbandoned == true case?

Reply:
Because we haven't cancelled the response body in that case. It's the cancellation of the response body that requires us to signal the waiter.